File: functions/connectome_slicer.R
# -----------------------------------------------------
# Slices a connectome matrix based on lookup tables
# -----------------------------------------------------
## Collect arguments
args <- commandArgs(TRUE)

## Default setting when no arguments passed
if(length(args) < 1) {
  args <- c("--help")
}

## Help section
if("--help" %in% args) {
  cat("
      micapipe connectome slicer

      Arguments:
      --conn=<path to connectivty matrix>   - character
      --lut1=<path to LUT>                  - character
      --lut2=<path to LUT>                  - character
      --mica=<MICAPIPE path>                - character
      --help                                - print this text

      Example:
      connectome_slicer.R --conn='HC10_10M_glasser-360_full-connectome.txt' --lut1='lut_subcortical-cerebellum_mics.csv' --lut2='lut_glasser-360_mics.csv' \n\n")

  q(save="no")
}

## Parse arguments (we expect the form --arg=value)
parseArgs <- function(x) strsplit(sub("^--", "", x), "=")
argsDF <- as.data.frame(do.call("rbind", parseArgs(args)))
argsL <- as.list(as.character(argsDF$V2))
names(argsL) <- argsDF$V1

exit <- function() {
  .Internal(.invokeRestart(list(NULL, NULL), NULL))
}
arg.miss <- function() {
  cat("---------------------------------------------------------
    ERROR ... Connectome slicer: A mandatory argument is missing
      ---------------------------------------------------------
      ")
  print(paste0("Connectome     : ",argsL$conn))
  print(paste0("LUT-1          : ",argsL$lut1))
  print(paste0("LUT-2          : ",argsL$lut2))
  print(paste0("MICAPIPE-path  : ",argsL$mica))
}

## Arg1 default
if(is.null(argsL$conn)) {
  arg.miss()
  exit()
}

## Arg2 default
if(is.null(argsL$lut1)) {
  arg.miss()
  exit()
}

## Arg3 default
if(is.null(argsL$lut2)) {
  arg.miss()
  exit()
}

## Arg3 default
if(is.null(argsL$mica)) {
  arg.miss()
  exit()
}

# -----------------------------------------------------

# Subcortical LUT
lut1 <- read.csv(argsL$lut1)

# Cortical LUT
lut2 <- read.csv(argsL$lut2)
indx <- sort(c(lut1$mics, lut2$mics))

# Connectivity matrix
# Read the matrix
conn <- argsL$conn
M <- as.matrix(read.table(conn, sep = " ",header = FALSE))
print(paste0("INFO.... Connectome dimensions: ",dim(M)[1]," x ",dim(M)[2]))
if (dim(M)[1]==length(indx)) { print("Connectome is already sliced!"); # exit()
} else {
  print(paste0("INFO.... Connectome new dimensions: ",length(indx)," x ",length(indx)))
  M <- M[indx, indx]
  # Overwrites the connectome sliced
  print(paste0("    Path=",conn))
  write.table(M, conn, sep = " ", row.names = FALSE, col.names = FALSE)
   }

# Loads the MICs colormaps
mica.dir <- paste0(argsL$mica,"/functions/")
load(file=paste0(mica.dir,"cmap_MICs.Rdata"))

# Save png for QC purposes
nom <-  gsub(".txt","", strsplit(conn, split = "connectomes/")[[1]][2])
qc <- paste0(strsplit(conn, split = "proc_dwi/")[[1]][1],"QC/png/")
print(paste0("INFO.... Saving Connectome as png for QC"))
print(paste0("    Path=",qc, nom, ".png"))
png(paste0(qc, nom, ".png"))
  # Mirror matrix completes inferior triangle
  M[lower.tri(M)] <- t(M)[lower.tri(M)]
  image(log(M), axes=FALSE, main=nom, col=cmap.SC(256))
dev.off()

File: functions/connectome_slicer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Generates PNG images from a nifti for Quality Check in micapipe


    Parameters
    ----------

    conn     : str, path to connectivity matrix

    lut1     : str, path to LUT

    lut2     : str, path to LUT

    mica   : str, MICAPIPE path

    Usage
    -----
    connectome_slicer.py --conn outname.png --lut1 --lut2 --mica

Created on June 2023 (the year of light)

@author: rcruces
"""

import argparse
import pandas as pd
import nibabel as nib
import numpy as np
import os

# Function save as gifti
def save_gii(data_array, file_name):
    # Initialize gifti: NIFTI_INTENT_SHAPE - 2005, FLOAT32 - 16
    gifti_data = nib.gifti.GiftiDataArray(data=data_array, intent=2005, datatype=16)

    # this is the GiftiImage class
    gifti_img = nib.gifti.GiftiImage(meta=None, darrays=[gifti_data])

    # Save the new GIFTI file
    nib.save(img=gifti_img, filename=file_name)

# Argument parsing
parser = argparse.ArgumentParser(description="micapipe connectome slicer")
parser.add_argument("--conn", help="path to connectivity matrix", required=True)
parser.add_argument("--lut1", help="path to LUT", required=True)
parser.add_argument("--lut2", help="path to LUT", required=True)
parser.add_argument("--mica", help="MICAPIPE path", required=True)
args = parser.parse_args()

# Read LUT files
lut1 = pd.read_csv(args.lut1)
lut2 = pd.read_csv(args.lut2)
indx = sorted(list(set(lut1['mics']).union(set(lut2['mics']))))

# Read connectivity matrix
M = pd.read_csv(args.conn, sep=" ", header=None).values
print("INFO.... Connectome dimensions: {} x {}".format(M.shape[0], M.shape[1]))

# Check if connectivity matrix contains "fsLR-5k"
if "fsLR-5k" in args.conn:
    print("Connectome is 'fsLR-5k'. Saving as GIFTI without slicing.")
    output_file = args.conn.replace("txt", "shape.gii")
    save_gii(M, output_file)
    print("GIFTI file saved as:", output_file)
else:
    if M.shape[0] == len(indx):
        print("Connectome is already sliced!")
    else:
        print("INFO.... Connectome new dimensions: {} x {}".format(len(indx), len(indx)))
        indx = [i - 1 for i in indx]
        M = M[np.ix_(indx, indx)]
        # Save the GIFTI data to a file
        output_file = args.conn.replace("txt", "shape.gii")
        save_gii(M, output_file)
        print("GIFTI file saved as:", output_file)

# Remove the txt file
os.remove(args.conn)
